### PR TITLE
Prune join columns

### PR DESCRIPTION
**Pull request by alandpost**
This patch adds three rules to the iterative optimizer for pruning the  outputs
and inputs of JoinNode.

Cross joins are handled separately, as they always pass all their child columns
through.  Thus, any project-off can be pushed down to below the cross join.

Non-cross joins pass through a subset of their child columns, so we need to have
a rule (PruneJoinChildrenColumns) to turn any projecting-off of child columns
into actual project nodes for the other rules to match against.  There's also a
rule to incorporate any projecting-off from a parent project node into the join
node itself (PruneJoinColumns).

https://github.com/prestodb/presto/issues/7154

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinColumns.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                                        .withExactOutputs("rightValue")));
+    }
+
+    @Test
+    public void testAllInputsReferenced()
+    {
+        tester.assertThat(new PruneJoinColumns())
+                .on(p -> buildProjectedJoin(p, symbol -> true))
+                .doesNotFire();
+    }
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinColumns.java]: Please add test ensuring the rule does not fire on `x-join` and maybe a test covering non-inner join too?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 2]
```diff
@@ -71,6 +71,7 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
                 parent.replaceChildren(
                         child.getSources())));
     }
+
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: Should be fixed in a39d73de31e3f2972820b68ba2232d5050f093bb (and original PR) instead.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 3]
```diff
@@ -121,4 +122,30 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
                         .collect(toImmutableSet())
         );
     }
+
+    /**
+     * @return The original node, with identity projections possibly inserted between node and each child, limiting the columns to those permitted.
+     * Returns a present Optional iff at least one child was rewritten.
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: Typo.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 4]
```diff
@@ -121,4 +122,30 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
...
+     */
+    public static Optional<PlanNode> restrictChildOutputs(PlanNodeIdAllocator idAllocator, PlanNode node, List<Set<Symbol>> permittedChildOutputs)
+    {
+        if (node.getSources().size() != permittedChildOutputs.size()) {
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: Please use `checkArgument` instead.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 2 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: I'm sorry, but I don't see what should be fixed.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 3 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: If you mean "iff", that's "if and only if"

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 5]
```diff
@@ -85,4 +89,36 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
...
                 Assignments.identity(restrictedOutputs)));
     }
+
+    public static List<Set<Symbol>> requiredJoinInputs(JoinNode node, Set<Symbol> requiredJoinOutputs)
+    {
+        Set<Symbol> globallyUsable = requiredJoinOutputs;
+        if (node.getFilter().isPresent()) {
+            globallyUsable = ImmutableSet.<Symbol>builder()
+                    .addAll(requiredJoinOutputs)
+                    .addAll(DependencyExtractor.extractUnique(node.getFilter().get()))
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: GroupReference is a PlanNode, not a Node -- so its filter can't be one, as I understand it.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinColumns.java (hunk 1 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinColumns.java]: I'll add the cross join case.  Except for whether the node is a cross join, the rule never looks at the join type field, so I don't think a non-inner-join test is worthwhile.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 3 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: My bad then. Didn't know that. Thanks for explanation. :)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 5 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: You are right.

**Comment by kokosing** [comment]
Please rebase.

**Comment by alandpost** [comment]
Rebased

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 6]
```diff
@@ -87,4 +91,35 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
...
+                .addAll(node.getLeftHashSymbol().map(Stream::of).orElse(Stream.empty()).iterator())
+                .build();
+
+        Set<Symbol> rightUsable = ImmutableSet.<Symbol>builder()
+                .addAll(globallyUsable)
+                .addAll(node.getCriteria().stream().map(JoinNode.EquiJoinClause::getRight).iterator())
+                .addAll(node.getRightHashSymbol().map(Stream::of).orElse(Stream.empty()).iterator())
+                .build();
+
+        return ImmutableList.of(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: I do not like the idea of returning a list here, lets create some class or have two methods one per each side of join

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneCrossJoinColumns.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        tester = new RuleTester();
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        closeAllRuntimeException(tester);
+        tester = null;
+    }
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneCrossJoinColumns.java]: [optional]
Maybe you could create some BaseRuleTest which could have a `tester`, `setUp` and `tearDown` inside.
That way it will be less boilerplate in all the rule tests.

I know it exceeds this PR, so if you don't want to do it, then I can do it in separate PR.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 8]
```diff
@@ -87,4 +91,35 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
                         node,
                         Assignments.identity(restrictedOutputs)));
     }
+
+    public static List<Set<Symbol>> requiredJoinInputs(JoinNode node, Set<Symbol> requiredJoinOutputs)
+    {
+        Set<Symbol> globallyUsable = requiredJoinOutputs;
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: I did it this way because below the if block, I wanted to use globallyUsable twice, once for the left child and once for the right.  If I just copy-paste the common code and make two functions, one for the left child and one for the right, then a single builder would work.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneCrossJoinColumns.java (hunk 7 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneCrossJoinColumns.java]: I'll make a separate PR.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 6 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: I figured out that the cross join pruning doesn't actually need to worry about this, so I can just inline the functions into the join child pruning.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.sql.planner.iterative.rule.Util.pruneInputs;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+/**
+ * Non-cross joins support output symbol selection, so absorb any project-off into the node.
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java]: I am not much a big fan of the fact that cross join requires special treatment. IMO cross join is like regular inner join but without any join condition, that way I think all the code which supports inner join should be able to support cross join as well. 

That way I would restrict outputs of join children in this rule, and cross join rule won't be needed.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinColumns.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        closeAllRuntimeException(tester);
+        tester = null;
+    }
+
+    private static PlanNode buildProjectedJoin(PlanBuilder p, Predicate<Symbol> projectionFilter)
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinColumns.java]: put this method at the bottom

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java [hunk 11]
```diff
@@ -63,8 +63,6 @@ public Pattern getPattern()
                 .map(ImmutableSet::copyOf)
                 .map(dependencies ->
                         parent.replaceChildren(ImmutableList.of(
-                                joinNode.replaceChildren(ImmutableList.of(
-                                        restrictOutputs(idAllocator, joinNode.getLeft(), dependencies).orElse(joinNode.getLeft()),
-                                        restrictOutputs(idAllocator, joinNode.getRight(), dependencies).orElse(joinNode.getRight()))))));
+                                restrictBinaryChildrenOutputs(idAllocator, joinNode, dependencies, dependencies).get())));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java]: I am not sure if previous version wasn't better. Maybe I got used to restrictOutputs, and so it was easier to me understand what it does.

Anyway, please extract this unrelated change as separate commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 12]
```diff
@@ -87,4 +88,54 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
...
+    public static Optional<PlanNode> restrictChildOutputs(PlanNodeIdAllocator idAllocator, PlanNode node, Set<Symbol> permittedChildOutputs)
+    {
+        return restrictChildrenOutputs(idAllocator, node, ImmutableList.of(permittedChildOutputs));
+    }
+
+    /**
+     * @return The original node, with identity projections possibly inserted between node and each child, limiting the columns to those permitted.
+     * Returns a present Optional iff at least one child was rewritten.
+     */
+    public static Optional<PlanNode> restrictBinaryChildrenOutputs(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: do you need emphases with `Binary` in name that it supports binary children nodes? IMO the fact it accepts two parameters is enough. 

Well I would go even one step further and create a method which accepts a var-args parameter or just a list, and then check if number of children matches to the number of parameters.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinChildrenColumns.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        closeAllRuntimeException(tester);
+        tester = null;
+    }
+
+    private static PlanNode buildJoin(PlanBuilder p, Predicate<Symbol> joinOutputFilter)
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinChildrenColumns.java]: please put this method at the bottom

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 14]
```diff
@@ -87,4 +88,54 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
                         node,
                         Assignments.identity(restrictedOutputs)));
     }
+
+    /**
+     * @return The original node, with identity projections possibly inserted between node and the child, limiting the columns to those permitted.
+     * Returns a present Optional iff the child was rewritten.
+     */
+    public static Optional<PlanNode> restrictChildOutputs(PlanNodeIdAllocator idAllocator, PlanNode node, Set<Symbol> permittedChildOutputs)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: is it used? If not please remove.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java (hunk 11 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java]: I thought the rule was to introduce two uses of a function in the commit that introduces that function, to justify its existence?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneCrossJoinColumns.java [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+    private static PlanNode buildProjectedCrossJoin(PlanBuilder p, Predicate<Symbol> projectionFilter)
```
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneCrossJoinColumns.java]: OK

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java (hunk 9 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java]: Cross join should not require special treatment, but does because of limitations in the evaluator which are enforced in assertions in the JoinNode constructor.  I tried to mention this in the commit comments.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinColumns.java (hunk 10 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinColumns.java]: OK

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 14 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: I'll do the varargs thing you requested, instead.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 12 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: I thought you wanted explicit handling of binary children, as opposed to lists, but I'll change it to be varargs.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinChildrenColumns.java (hunk 13 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneJoinChildrenColumns.java]: OK

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java [hunk 16]
```diff
@@ -95,6 +95,9 @@ public JoinNode(@JsonProperty("id") PlanNodeId id,
                 .build();
         checkArgument(inputSymbols.containsAll(outputSymbols), "Left and right join inputs do not contain all output symbols");
         checkArgument(!isCrossJoin() || inputSymbols.equals(outputSymbols), "Cross join does not support output symbols pruning or reordering");
+
+        checkArgument((!criteria.isEmpty()) || !leftHashSymbol.isPresent(), "Left hash symbol is only valid in an equijoin");
+        checkArgument((!criteria.isEmpty()) || !rightHashSymbol.isPresent(), "Right hash symbol is only valid in an equijoin");
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java]: By this, I assume that you're asking for:
`checkArgument(!(criteria.isEmpty() && rightHashSymbols.isPresent()), ... );`
I will do that.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java (hunk 11 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java]: Anyway, I'll move it to a separate commit within this PR.

**Comment by alandpost** [comment]
@kokosing Please take another look.  I believe I've incorporated all your suggestions.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 17]
```diff
@@ -127,6 +128,7 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
 
         // TODO: Once we've migrated handling all the plan node types, replace uses of PruneUnreferencedOutputs with an IterativeOptimizer containing these rules.
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: That can be a separate PR, once we get the outstanding PRs committed.  There's no rush.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+            return Optional.empty();
+        }
+
+        JoinNode joinNode = (JoinNode) child;
+        if (!joinNode.isCrossJoin()) {
+            return Optional.empty();
+        }
+
+        return pruneInputs(child.getOutputSymbols(), parent.getAssignments().getExpressions())
+                .map(ImmutableSet::copyOf)
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java]: I agree that it is strange that it returns a List.  Martin wrote it that way, and I haven't changed it.  I'll do that in a future PR.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java (hunk 9 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java]:
```
checkArgument(!isCrossJoin() || inputSymbols.equals(outputSymbols), "Cross join does not support output symbols pruning or reordering");
```

So if you restrict outputs of the children of join and in the same time you limit outputSymbols of this join properly, then this check should pass.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java [hunk 19]
```diff
@@ -63,8 +63,6 @@ public Pattern getPattern()
                 .map(ImmutableSet::copyOf)
                 .map(dependencies ->
                         parent.replaceChildren(ImmutableList.of(
-                                joinNode.replaceChildren(ImmutableList.of(
-                                        restrictOutputs(idAllocator, joinNode.getLeft(), dependencies).orElse(joinNode.getLeft()),
-                                        restrictOutputs(idAllocator, joinNode.getRight(), dependencies).orElse(joinNode.getRight()))))));
+                                restrictChildOutputs(idAllocator, joinNode, dependencies, dependencies).get())));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java]: what if restrictChildOutputs return an `Optional.empty()`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java (hunk 9 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java]: There needs to be two rules for non-cross join, one that looks for a project over a join, and one that looks for a join, because the join can itself project off columns.  Currently, the first rule doesn't need to include the logic to find which join inputs are needed, so that logic can be included inline in the second rule.  If the first rule also prunes the join inputs, then the logic would need to be factored out.  I think a separate rule for cross join is cleaner, and matches a future in which NestedLoopJoin is a different plan node than HashJoin, as opposed the current situation where the two operators share a plan node that doesn't actually fit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java (hunk 19 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCrossJoinColumns.java]: It won't, because cross joins pass through all their inputs.  So one of the children will have been pruned, if the outputs of the join can be pruned.

**Comment by alandpost** [comment]
@kokosing I rebased to accommodate the extractor class rename.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java (hunk 9 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java]: > There needs to be two rules for non-cross join, one that looks for a project over a join, and one that looks for a join, because the join can itself project off columns. Currently, the first rule doesn't need to include the logic to find which join inputs are needed, so that logic can be included inline in the second rule. If the first rule also prunes the join inputs, then the logic would need to be factored out.

This is clear to me, and you already explained that me a couple of times ;)

> I think a separate rule for cross join is cleaner

That is your opinion 

> and matches a future in which NestedLoopJoin is a different plan node than HashJoin, as opposed the current situation where the two operators share a plan node that doesn't actually fit.

IMO NestedLoopJoin, HashJoin, MergeJoin any other join are physical plan nodes, while JoinNode is logical. I would rather expect to have your rules to work on logical nodes. In such case nothing should change. 

Anyway I see you have a strong opinion about this. So let it be as you want. I will create a PR in which I will check my idea and give it to you for a review. Not because I want so much to persuade you, but to just give it a try ;)
